File: v4/rust/iprange-livedb/src/work.rs
//! Test-only accounting for necessary low-level work.

#[cfg(test)]
use std::cell::RefCell;

#[cfg(test)]
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub(crate) struct Snapshot {
    pub(crate) tree_lookups: u64,
    pub(crate) tree_descents: u64,
    pub(crate) pages_visited: u64,
    pub(crate) page_parses: u64,
    pub(crate) key_probes: u64,
    pub(crate) cell_probes: u64,
    pub(crate) leaf_validations: u64,
    pub(crate) bitmap_probes: u64,
    pub(crate) slot_reads: u64,
    pub(crate) slot_scan_steps: u64,
    pub(crate) edit_fit_probes: u64,
    pub(crate) pages_created: u64,
    pub(crate) pages_copied: u64,
    pub(crate) pages_split: u64,
    pub(crate) first_fence_updates: u64,
    pub(crate) edge_path_checks: u64,
    pub(crate) leaf_locator_hits: u64,
    pub(crate) leaf_locator_misses: u64,
    pub(crate) leaf_locator_fallbacks: u64,
    pub(crate) pages_retired: u64,
    pub(crate) pages_reclaimed: u64,
    pub(crate) pages_sealed: u64,
    pub(crate) ranges_consumed: u64,
    pub(crate) ranges_emitted: u64,
    pub(crate) ranges_split: u64,
    pub(crate) ranges_coalesced: u64,
    pub(crate) catalog_lookups: u64,
    pub(crate) catalog_interns: u64,
    pub(crate) membership_lookups: u64,
    pub(crate) membership_interns: u64,
    pub(crate) structure_lookups: u64,
    pub(crate) structure_interns: u64,
    pub(crate) structure_decodes: u64,
    pub(crate) mapping_growths: u64,
    pub(crate) mapping_remaps: u64,
    pub(crate) mapping_flushes: u64,
    pub(crate) file_syncs: u64,
    pub(crate) bytes_moved: u64,
    pub(crate) bytes_zeroed: u64,
    pub(crate) membership_leaf_reads: u64,
    pub(crate) membership_refcount_batches: u64,
    pub(crate) membership_delta_spills: u64,
    pub(crate) source_passes: u64,
    pub(crate) input_source_passes: u64,
    pub(crate) output_passes: u64,
    pub(crate) history_window_tests: u64,
    pub(crate) membership_decodes: u64,
    pub(crate) membership_decode_cache_hits: u64,
    pub(crate) membership_word_reads: u64,
    pub(crate) membership_combinations: u64,
    pub(crate) membership_intern_cache_hits: u64,
    pub(crate) aggregation_contributions: u64,
    pub(crate) aggregation_results: u64,
    pub(crate) join_advances: u64,
}

#[cfg(test)]
thread_local! {
    static CURRENT: RefCell<Snapshot> = RefCell::new(Snapshot::default());
}

#[cfg(test)]
fn update(change: impl FnOnce(&mut Snapshot)) {
    CURRENT.with_borrow_mut(change);
}

macro_rules! event {
    ($function:ident, $field:ident) => {
        #[inline(always)]
        pub(crate) fn $function(amount: u64) {
            #[cfg(test)]
            update(|current| current.$field += amount);
            #[cfg(not(test))]
            let _ = amount;
        }
    };
}

event!(tree_lookup, tree_lookups);
event!(tree_descent, tree_descents);
event!(page_visited, pages_visited);
event!(page_parse, page_parses);
event!(key_probe, key_probes);
event!(cell_probe, cell_probes);
event!(leaf_validation, leaf_validations);
event!(bitmap_probe, bitmap_probes);
event!(slot_read, slot_reads);
event!(slot_scan_step, slot_scan_steps);
event!(edit_fit_probe, edit_fit_probes);
event!(page_created, pages_created);
event!(page_copied, pages_copied);
event!(page_split, pages_split);
event!(first_fence_update, first_fence_updates);
event!(edge_path_check, edge_path_checks);
event!(leaf_locator_hit, leaf_locator_hits);
event!(leaf_locator_miss, leaf_locator_misses);
event!(leaf_locator_fallback, leaf_locator_fallbacks);
event!(page_retired, pages_retired);
event!(page_reclaimed, pages_reclaimed);
event!(page_sealed, pages_sealed);
event!(range_consumed, ranges_consumed);
event!(range_emitted, ranges_emitted);
event!(range_split, ranges_split);
event!(range_coalesced, ranges_coalesced);
event!(catalog_lookup, catalog_lookups);
event!(catalog_intern, catalog_interns);
event!(membership_lookup, membership_lookups);
event!(membership_intern, membership_interns);
event!(structure_lookup, structure_lookups);
event!(structure_intern, structure_interns);
event!(structure_decode, structure_decodes);
event!(mapping_growth, mapping_growths);
event!(mapping_remap, mapping_remaps);
event!(mapping_flush, mapping_flushes);
event!(file_sync, file_syncs);
event!(bytes_moved, bytes_moved);
event!(bytes_zeroed, bytes_zeroed);
event!(membership_leaf_read, membership_leaf_reads);
event!(membership_refcount_batch, membership_refcount_batches);
event!(membership_delta_spill, membership_delta_spills);
event!(source_pass, source_passes);
event!(input_source_pass, input_source_passes);
event!(output_pass, output_passes);
event!(history_window_test, history_window_tests);
event!(membership_decode, membership_decodes);
event!(membership_decode_cache_hit, membership_decode_cache_hits);
event!(membership_word_read, membership_word_reads);
event!(membership_combination, membership_combinations);
event!(membership_intern_cache_hit, membership_intern_cache_hits);
event!(aggregation_contribution, aggregation_contributions);
event!(aggregation_result, aggregation_results);
event!(join_advance, join_advances);

#[cfg(test)]
pub(crate) fn reset() {
    CURRENT.with_borrow_mut(|current| *current = Snapshot::default());
}

#[cfg(test)]
pub(crate) fn snapshot() -> Snapshot {
    CURRENT.with_borrow(|current| *current)
}

#[cfg(test)]
pub(crate) fn measure<T>(operation: impl FnOnce() -> T) -> (T, Snapshot) {
    reset();
    let result = operation();
    (result, snapshot())
}

File: v4/go/internal/work/work.go
//go:build !v4work

// Package work records test-only necessary-work counters for the reader and
// writer cores. Every counter call compiles to an inlineable no-op in
// production builds (Enabled is const false); build the module with -tags
// v4work to enable the counters and the snapshot API. Tests pin the exact
// counts so a change that adds or removes required work on the hot path is
// visible (mirroring v4/rust/iprange-livedb/src/work.rs for the Go peer).
package work

// Enabled reports whether the counters are compiled in.
const Enabled = false

// TreeLookup counts one root-to-leaf query (range, catalog, membership
// dictionary, blob leaf, structure table).
func TreeLookup(uint64) {}

// TreeDescent counts one branch-follow during a tree walk.
func TreeDescent(uint64) {}

// PageVisit counts one visited page view on the reader hot paths.
func PageVisit(uint64) {}

// PageParse counts one page-header decode. The reader page owner decodes
// exactly one header per visited page, so on the reader hot paths PageParse
// moves with PageVisit and the pinned reader tests treat them as one
// invariant. Writer edit paths decode headers through their own parse owner
// and can re-inspect a page (COW first-touch, cell apply), so there
// PageParse is not tied to PageVisit.
func PageParse(uint64) {}

// KeyProbe counts one key-only comparison during a fixed-tree search.
func KeyProbe(uint64) {}

// LeafValidation counts one decode of a selected leaf record.
func LeafValidation(uint64) {}

// WordRead counts one 8-byte membership bitmap word read.
func WordRead(uint64) {}

// MembershipDecode counts one point-match membership decode (one whole
// matching-feeds scan; Rust work::membership_decode parity).
func MembershipDecode(uint64) {}

// StructureDecode counts one structured payload decode.
func StructureDecode(uint64) {}

// StructureIntern counts one structured payload intern (Rust
// work::structure_intern parity).
func StructureIntern(uint64) {}

// MappingRemap counts one actual mapping resize (remap or shrink that
// re-establishes the mapping); same-size no-ops do not count.
func MappingRemap(uint64) {}

// MappingGrowth counts one file-and-mapping extension.
func MappingGrowth(uint64) {}

// MappingFlush counts one whole-mapping msync (MS_SYNC).
func MappingFlush(uint64) {}

// FileSync counts one stable-storage file sync (fsync / F_FULLFSYNC).
func FileSync(uint64) {}

// CellProbe counts one per-record decode attempt during slotted-page
// inspection (fixed cells and variable records).
func CellProbe(uint64) {}

// SlotRead counts one persistent slot-array value read.
func SlotRead(uint64) {}

// SlotScanStep counts one slot visited during a layout scan or adjustment.
func SlotScanStep(uint64) {}

// EditFitProbe counts one fit pre-check before a structural edit.
func EditFitProbe(uint64) {}

// BitmapProbe counts one bitmap word or summary read during a bitmap walk.
func BitmapProbe(uint64) {}

// PageCreated counts one page allocated (private or tail) for a draft.
func PageCreated(uint64) {}

// PageCopied counts one complete page copied for COW (copy_for_cow).
func PageCopied(uint64) {}

// PageSplit counts one page split during a structural edit.
func PageSplit(uint64) {}

// PageRetired counts one page retired to the retirement tree.
func PageRetired(uint64) {}

// PageReclaimed counts one retired page extent reclaimed into the free
// bitmap.
func PageReclaimed(uint64) {}

// PageSealed counts one page whose checksum was stamped.
func PageSealed(uint64) {}

// BytesMoved counts bytes copied within or across mapped pages.
func BytesMoved(uint64) {}

// BytesZeroed counts bytes zeroed inside mapped pages.
func BytesZeroed(uint64) {}

// FirstFenceUpdate counts one branch first-key (fence) update during a
// split or delete propagation.
func FirstFenceUpdate(uint64) {}

// EdgePathCheck counts one cached-edge position verification during a
// monotonic edge insertion (Rust edge_path_check).
func EdgePathCheck(uint64) {}

// RangeConsumed counts one range record read by a reader cursor (Rust
// range_consumed; written records are counted separately by
// RangeEmitted).
func RangeConsumed(uint64) {}

// RangeEmitted counts one range record written during a range edit.
func RangeEmitted(uint64) {}

// CatalogLookup counts one feed-name or feed-index catalog lookup (Rust
// work::catalog_lookup).
func CatalogLookup(uint64) {}

// RangeSplit counts one range record split into two during a rewrite.
func RangeSplit(uint64) {}

// RangeCoalesced counts one adjacency merge of two same-value ranges.
func RangeCoalesced(uint64) {}

// InputSourcePass counts one logical pass over one input source (Rust
// work::input_source_pass: one for a membership aggregation scan, two for
// a two-source join).
func InputSourcePass(uint64) {}

// MembershipDecodeCacheHit counts one selected-membership sequence served
// from the bounded decode cache (Rust work::membership_decode_cache_hit).
func MembershipDecodeCacheHit(uint64) {}

// MembershipWordRead counts one membership bitmap word read by the
// selected-membership decoder (Rust work::membership_word_read).
func MembershipWordRead(uint64) {}

// AggregationContribution counts one exact contribution folded into a
// feed total, pair total, cross cell, or uncovered total (Rust
// work::aggregation_contribution).
func AggregationContribution(uint64) {}

// AggregationResult counts one emitted aggregation or join result record
// (Rust work::aggregation_result).
func AggregationResult(uint64) {}

// JoinAdvance counts one sweep step of an ordered join (Rust
// work::join_advance).
func JoinAdvance(uint64) {}

// CatalogIntern counts one feed catalog internment (Rust
// work::catalog_intern).
func CatalogIntern(uint64) {}

// MembershipLookup counts one membership dictionary lookup (Rust
// work::membership_lookup).
func MembershipLookup(uint64) {}

// MembershipIntern counts one membership dictionary intern attempt (Rust
// work::membership_intern).
func MembershipIntern(uint64) {}

// MembershipRefcountBatch counts one applied dictionary refcount batch
// (Rust work::membership_refcount_batch).
func MembershipRefcountBatch(uint64) {}

// OutputPass counts one whole output-build pass over an input source
// (Rust work::output_pass).
func OutputPass(uint64) {}

// MembershipInternCacheHit counts one membership intern served from the
// bounded sequence cache (Rust work::membership_intern_cache_hit).
func MembershipInternCacheHit(uint64) {}

// SourcePass counts one logical pass over one history source (Rust
// work::source_pass).
func SourcePass(uint64) {}

// HistoryWindowTest counts one per-window before/after test during a
// history projection (Rust work::history_window_test).
func HistoryWindowTest(uint64) {}

// MembershipCombination counts one membership bitmap combination
// (Rust work::membership_combination).
func MembershipCombination(uint64) {}

// MembershipDeltaSpill counts one refcount delta spilled from the
// pending slots into the delta tree (Rust work::membership_delta_spill).
func MembershipDeltaSpill(uint64) {}

// LeafLocatorHit counts one private input served directly by the leaf
// locator cache (Rust work::leaf_locator_hit).
func LeafLocatorHit(uint64) {}

// LeafLocatorMiss counts one private input with no cached leaf (Rust
// work::leaf_locator_miss).
func LeafLocatorMiss(uint64) {}

// LeafLocatorFallback counts one private input whose cached leaf did not
// fit the gap and fell back to a full descent (Rust
// work::leaf_locator_fallback).
func LeafLocatorFallback(uint64) {}
